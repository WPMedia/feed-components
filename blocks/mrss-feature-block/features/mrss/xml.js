import PropTypes from 'fusion:prop-types'
import Consumer from 'fusion:consumer'
import get from 'lodash/get'
import moment from 'moment'
import getProperties from 'fusion:properties'
import { resizerKey } from 'fusion:environment'
import { BuildContent } from '@wpmedia/feeds-content-elements'
import { generatePropsForFeed } from '@wpmedia/feeds-prop-types'
import { buildResizerURL } from '@wpmedia/feeds-resizer'
import { findVideo } from '@wpmedia/feeds-find-video-stream'

const jmespath = require('jmespath')

const rssTemplate = (
  elements,
  {
    channelTitle,
    channelDescription,
    channelPath,
    channelCopyright,
    channelTTL,
    channelUpdatePeriod,
    channelUpdateFrequency,
    channelCategory,
    channelLogo,
    imageTitle,
    imageCaption,
    imageCredits,
    itemTitle,
    itemDescription,
    pubDate,
    itemCredits,
    itemCategory,
    includeContent,
    resizerURL,
    domain,
    feedTitle,
    feedLanguage,
    videoInfo,
    mrssBuildContent,
  },
) => ({
  rss: {
    '@xmlns:atom': 'http://www.w3.org/2005/Atom',
    '@version': '2.0',
    '@xmlns:media': 'http://search.yahoo.com/mrss/',
    channel: {
      title: `${channelTitle || feedTitle + ' Videos'}`,
      link: domain,
      'atom:link': {
        '@href': `${domain}${channelPath}`,
        '@rel': 'self',
        '@type': 'application/rss+xml',
      },
      description: channelDescription || feedTitle,
      pubDate: moment.utc(Date.now()).format('ddd, DD MMM YYYY HH:mm:ss ZZ'),
      ...(channelLogo && {
        image: {
          url: buildResizerURL(channelLogo, resizerKey, resizerURL),
          title: `${channelTitle || feedTitle + ' Videos'}`,
          link: domain,
        },
      }),

      item: elements.map((s) => {
        const url = `${domain}${s.website_url || s.canonical_url}`
        const img =
          s.promo_items && (s.promo_items.basic || s.promo_items.lead_art)
        const videoStream = findVideo(s, videoInfo)

        return {
          title: `${jmespath.search(s, itemTitle)}`,
          link: url,
          ...(itemCredits &&
            (jmespath.search(img, itemCredits) || []).length && {
              'media:credit': {
                '@role': 'producer',
                $: jmespath.search(img, itemCredits).join(','),
              },
            }),
          ...(itemDescription && {
            description: jmespath.search(s, itemDescription),
          }),
          guid: {
            '@isPermaLink': false,
            '#': url,
          },
          referenceid: s._id,
          pubDate: moment
            .utc(s[pubDate])
            .format('ddd, DD MMM YYYY HH:mm:ss ZZ'),

          'media:content': {
            '@': {
              isDefault: 'true',
              ...(s.duration && {
                duration: Math.trunc(s.duration / 1000),
              }),
              ...(videoStream && {
                ...(videoInfo.url && { url: videoInfo.url }),
                ...(videoInfo.height && { height: videoInfo.height }),
                ...(videoInfo.width && { width: videoInfo.width }),
                ...(videoInfo.bitrate && { bitrate: videoInfo.bitrate }),
                ...(videoInfo.stream_type && {
                  type: videoInfo.stream_type,
                  '#': videoStream,
                }),
              }),
            },
          },
          'media:keywords': (
            jmespath.search(s, 'taxonomy.seo_keywords[*]') || []
          ).join(','),
          ...(s.description ||
            (s.subheadlines &&
              s.subheadlines.basic && {
                'media:caption': {
                  $: s.description || s.subheadlines.basic,
                },
              })),
          'media:transcript': s.transcript,
          ...(s.taxonomy &&
            s.taxonomy.primary_section &&
            s.taxonomy.primary_section.name && {
              'media:category': s.taxonomy.primary_section.name,
            }),
          ...(img &&
            img.url && {
              'media:thumbnail': {
                '@url': buildResizerURL(img.url, resizerKey, resizerURL),
              },
            }),
        }
      }),
    },
  },
})

export function Mrss({ globalContent, customFields, arcSite }) {
  const {
    resizerURL = '',
    feedDomainURL = '',
    feedTitle = '',
    feedLanguage = '',
  } = getProperties(arcSite)

  function MrssBuildContent() {
    BuildContent.call(this)
  }
  const mrssBuildContent = new MrssBuildContent()

  // can't return null for xml return type, must return valid xml template
  return rssTemplate(get(globalContent, 'content_elements', []), {
    ...customFields,
    resizerURL,
    domain: feedDomainURL,
    feedTitle,
    feedLanguage,
    mrssBuildContent,
  })
}

Mrss.propTypes = {
  customFields: PropTypes.shape({
    channelPath: PropTypes.string.tag({
      label: 'Path',
      group: 'Channel',
      description:
        'Path to the feed, excluding the domain, defaults to /arcio/mrss',
      defaultValue: '/arcio/mrss/',
    }),
    selectVideo: PropTypes.kvp.tag({
      label: 'Select video using',
      group: 'Video',
      description:
        'This criteria is used to filter videos encoded in the streams array',
      defaultValue: {
        height: '',
        width: '',
        bitrate: 5400,
        stream_type: 'mp4',
      },
    }),
    ...generatePropsForFeed('rss', PropTypes),
  }),
}
Mrss.label = 'MRSS'
export default Consumer(Mrss)
