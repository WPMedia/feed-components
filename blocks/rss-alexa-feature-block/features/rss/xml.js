import PropTypes from 'fusion:prop-types'
import Consumer from 'fusion:consumer'
import get from 'lodash/get'
import moment from 'moment'
import getProperties from 'fusion:properties'
import { resizerKey } from 'fusion:environment'
import { BuildContent } from '@wpmedia/feeds-content-elements'
import { generatePropsForFeed } from '@wpmedia/feeds-prop-types'
import { buildResizerURL } from '@wpmedia/feeds-resizer'

const jmespath = require('jmespath')

const cheerio = require('cheerio')

const getURLType = (url) => {
  const imgRegex = new RegExp(/(jpe?g|png|webp|PNG|JPG|WEBP)$/);
  const audioRegex = new RegExp('^(https?|ftp|file):\/\/(www.)?(.*?)\.(mp3)$');
  const videoRegex = new RegExp('^(https?|ftp|file):\/\/(www.)?(.*?)\.(mp4|ts)$');
  const uri = url.split(".");
  const type = uri[uri.length -1];
  const defaultType =  {label: 'Audio',type: `audio/${type}`};
  if(imgRegex.test(url)) {
    return {label: 'image',type: `image/${type}`};
  }
  if(audioRegex.test(url)){
    return {label: 'Audio',type: `audio/${type}`};
  }
  if(videoRegex.test(url)){
    return {label: 'Video',type: `video/${type}`};
  }
  
  return defaultType;
}

const rssTemplate = (
  elements,
  {
    audioAvailable,
    channelTitle,
    channelDescription,
    channelCopyright,
    channelTTL,
    channelCategory,
    channelLogo,
    itemTitle,
    pubDate,
    itemCategory,
    includeContent,
    resizerURL,
    resizerWidth,
    resizerHeight,
    domain,
    feedTitle,
    feedLanguage,
    rssBuildContent,
  },
) => ({
  rss: {
    '@version': '2.0',
    channel: {
      ...(channelTitle && { title: channelTitle }),
      link: `${domain}`,
      ...(channelDescription && { description: channelDescription }),
      lastBuildDate: moment
        .utc(new Date())
        .format('ddd, DD MMM YYYY HH:mm:ss ZZ'),
      ...(feedLanguage && { language: feedLanguage }),
      ...(channelCategory && { category: channelCategory }),
      ...(channelCopyright && {
        copyright: channelCopyright,
      }), // TODO Add default logic
      ...(channelTTL && { ttl: channelTTL }),

      ...(channelLogo && {
        image: {
          url: buildResizerURL(channelLogo, resizerKey, resizerURL),
          title: channelTitle || feedTitle,
          link: domain,
        },
      }),

      item: elements.map((s) => {
        let body, category,
        enclosureurl = jmespath.search(s, audioAvailable);
        const url = `${domain}${s.website_url || s.canonical_url || ''}`
        return {
          title: jmespath.search(s, itemTitle) || '',
          link: url,
          guid: s._id,
          pubDate: s[pubDate],
          ...(enclosureurl  &&
          {
            enclosure: {
              '@url': enclosureurl,
              '@label':  getURLType(enclosureurl).label,
              '@group': getURLType(enclosureurl).label,
              '@type': getURLType(enclosureurl).type
            }
          }),
          ...(itemCategory &&
            (category = jmespath.search(s, itemCategory)) &&
            category && { category: category }),
            description: cheerio
            .load(
              rssBuildContent.parse(
                s.content_elements || [],
                includeContent,
                domain,
                resizerKey,
                resizerURL,
                resizerWidth,
                resizerHeight,
              ) || '',
            )
            .text()
        }
      }),
    },
  },
})

export function Rss({ globalContent, customFields, arcSite }) {
  const {
    resizerURL = '',
    feedDomainURL = '',
    feedTitle = '',
    feedLanguage = '',
  } = getProperties(arcSite)

  const { width = 0, height = 0 } = customFields.resizerKVP || {};
  

  const rssBuildContent = new BuildContent()
 
  // can't return null for xml return type, must return valid xml template
  return rssTemplate(get(globalContent, 'content_elements', []), {
    ...customFields,
    resizerURL,
    resizerWidth: width,
    resizerHeight: height,
    domain: feedDomainURL,
    feedTitle,
    feedLanguage,
    rssBuildContent,
  })
}

Rss.propTypes = {
  customFields: PropTypes.shape({
    audioAvailable: PropTypes.string.tag({
      label: 'Audio',
      group: 'Audio',
      description: 'description',
      defaultValue: `content_elements[?type=='audio'].streams[].url|[0]`,
    }),
    ...generatePropsForFeed('rss', PropTypes, ['featuredImage', 'includePromo']),
  }),
}
Rss.label = 'RSS Alexa'
export default Consumer(Rss)
