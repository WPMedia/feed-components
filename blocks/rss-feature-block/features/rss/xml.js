import PropTypes from 'fusion:prop-types'
import Consumer from 'fusion:consumer'
import get from 'lodash/get'
import moment from 'moment'
import getProperties from 'fusion:properties'
import { resizerKey } from 'fusion:environment'
import { BuildContent } from '@wpmedia/feeds-content-elements'
import { generatePropsForFeed } from '@wpmedia/feeds-prop-types'
import { buildResizerURL } from '@wpmedia/feeds-resizer'

const jmespath = require('jmespath')

const rssTemplate = (
  elements,
  {
    channelTitle,
    channelDescription,
    channelPath,
    channelCopyright,
    channelTTL,
    channelUpdatePeriod,
    channelUpdateFrequency,
    channelCategory,
    channelLogo,
    imageTitle,
    imageCaption,
    imageCredits,
    itemTitle,
    itemDescription,
    pubDate,
    itemCategory,
    includePromo,
    includeContent,
    resizerURL,
    domain,
    feedTitle,
    feedLanguage,
    rssBuildContent,
  },
) => ({
  rss: {
    '@xmlns:atom': 'http://www.w3.org/2005/Atom',
    '@xmlns:content': 'http://purl.org/rss/1.0/modules/content/',
    '@xmlns:dc': 'http://purl.org/dc/elements/1.1/',
    '@xmlns:sy': 'http://purl.org/rss/1.0/modules/syndication/',
    '@version': '2.0',
    ...(includePromo && {
      '@xmlns:media': 'http://search.yahoo.com/mrss/',
    }),
    channel: {
      title: `${channelTitle || feedTitle}`,
      link: `${domain}`,
      'atom:link': {
        '@href': `${domain}${channelPath}`,
        '@rel': 'self',
        '@type': 'application/rss+xml',
      },
      description: `${channelDescription || feedTitle + ' News Feed'}`,
      lastBuildDate: moment
        .utc(new Date())
        .format('ddd, DD MMM YYYY HH:mm:ss ZZ'),
      ...(feedLanguage && { language: feedLanguage }),
      ...(channelCategory && { category: channelCategory }),
      ...(channelCopyright && {
        copyright: channelCopyright,
      }), // TODO Add default logic
      ...(channelTTL && { ttl: channelTTL }),
      ...(channelUpdatePeriod && {
        'sy:updatePeriod': channelUpdatePeriod,
      }),
      ...(channelUpdateFrequency && {
        'sy:updateFrequency': channelUpdateFrequency,
      }),
      ...(channelLogo && {
        image: {
          url: buildResizerURL(channelLogo, resizerKey, resizerURL),
          title: `${channelTitle || feedTitle}`,
          link: `${domain}`,
        },
      }),

      item: elements.map((s) => {
        let author, body, category
        const url = `${domain}${s.website_url || s.canonical_url}`
        const img =
          s.promo_items && (s.promo_items.basic || s.promo_items.lead_art)
        return {
          title: `${jmespath.search(s, itemTitle)}`,
          link: url,
          guid: {
            '#': url,
            '@isPermaLink': true,
          },
          ...((author = jmespath.search(s, 'credits.by[].name')) &&
            author && {
              'dc:creator': author.join(','),
            }),
          description: { $: jmespath.search(s, itemDescription) },
          pubDate: moment
            .utc(s[pubDate])
            .format('ddd, DD MMM YYYY HH:mm:ss ZZ'),
          ...(itemCategory &&
            (category = jmespath.search(s, itemCategory)) &&
            category && { category: category }),
          ...(includeContent !== '0' &&
            (body = rssBuildContent.parse(
              s.content_elements,
              includeContent,
              domain,
              resizerKey,
              resizerURL,
            )) &&
            body && {
              'content:encoded': {
                $: body,
              },
            }),
          ...(includePromo &&
            img &&
            img.url && {
              'media:content': {
                '@type': 'image/jpeg',
                '@url': buildResizerURL(img.url, resizerKey, resizerURL),
                ...(jmespath.search(img, imageCaption) && {
                  'media:description': {
                    '@type': 'plain',
                    $: jmespath.search(img, imageCaption),
                  },
                }),
                ...(jmespath.search(img, imageTitle) && {
                  'media:title': {
                    $: jmespath.search(img, imageTitle),
                  },
                }),
                ...((jmespath.search(img, imageCredits) || []).length && {
                  'media:credit': {
                    '@role': 'author',
                    '@scheme': 'urn:ebu',
                    '#': jmespath.search(img, imageCredits).join(','),
                  },
                }),
              },
            }),
        }
      }),
    },
  },
})

export function Rss({ globalContent, customFields, arcSite }) {
  const {
    resizerURL = '',
    feedDomainURL = '',
    feedTitle = '',
    feedLanguage = '',
  } = getProperties(arcSite)

  const rssBuildContent = new BuildContent()

  // can't return null for xml return type, must return valid xml template
  return rssTemplate(get(globalContent, 'content_elements', []), {
    ...customFields,
    resizerURL,
    domain: feedDomainURL,
    feedTitle,
    feedLanguage,
    rssBuildContent,
  })
}

Rss.propTypes = {
  customFields: PropTypes.shape({
    channelPath: PropTypes.string.tag({
      label: 'Path',
      group: 'Channel',
      description:
        'Path to the feed, excluding the domain, defaults to /arcio/rss',
      defaultValue: '/arcio/rss/',
    }),
    ...generatePropsForFeed('rss', PropTypes),
  }),
}
Rss.label = 'Standard RSS'
export default Consumer(Rss)
