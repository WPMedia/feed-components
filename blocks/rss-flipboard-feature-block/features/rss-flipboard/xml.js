import PropTypes from 'fusion:prop-types'
import Consumer from 'fusion:consumer'
import get from 'lodash/get'
import moment from 'moment'
import getProperties from 'fusion:properties'
import { resizerKey } from 'fusion:environment'
import { BuildContent } from '@wpmedia/feeds-content-elements'
import { generatePropsForFeed } from '@wpmedia/feeds-prop-types'
import { buildResizerURL } from '@wpmedia/feeds-resizer'

const jmespath = require('jmespath')

const rssTemplate = (
  elements,
  {
    channelTitle,
    channelDescription,
    channelPath,
    channelCopyright,
    channelTTL,
    channelUpdatePeriod,
    channelUpdateFrequency,
    channelCategory,
    channelLogo,
    imageTitle,
    imageCaption,
    imageCredits,
    itemTitle,
    itemDescription,
    pubDate,
    itemCredits,
    itemCategory,
    includeContent,
    resizerURL,
    resizerWidth,
    resizerHeight,
    domain,
    feedTitle,
    feedLanguage,
    flipboardBuildContent,
  },
) => ({
  rss: {
    '@xmlns:atom': 'http://www.w3.org/2005/Atom',
    '@xmlns:content': 'http://purl.org/rss/1.0/modules/content/',
    '@xmlns:dc': 'http://purl.org/dc/elements/1.1/',
    '@xmlns:sy': 'http://purl.org/rss/1.0/modules/syndication/',
    '@xmlns:media': 'http://search.yahoo.com/mrss/',
    '@version': '2.0',
    channel: {
      title: { $: channelTitle || feedTitle },
      link: domain,
      'atom:link': {
        '@href': `${domain}${channelPath}`,
        '@rel': 'self',
        '@type': 'application/rss+xml',
      },
      description: { $: channelDescription || `${feedTitle} News Feed` },
      ...(feedLanguage && { language: feedLanguage }),
      lastBuildDate: moment
        .utc(new Date())
        .format('ddd, DD MMM YYYY HH:mm:ss ZZ'),
      ...(channelCategory && { category: channelCategory }),
      ...(channelCopyright && {
        copyright: channelCopyright,
      }), // TODO Add default logic
      ...(channelTTL && { ttl: channelTTL }),
      ...(channelUpdatePeriod && {
        'sy:updatePeriod': channelUpdatePeriod,
      }),
      ...(channelUpdateFrequency && {
        'sy:updateFrequency': channelUpdateFrequency,
      }),
      ...(channelLogo && {
        image: {
          url: buildResizerURL(channelLogo, resizerKey, resizerURL),
          title: channelTitle || feedTitle,
          link: domain,
        },
      }),

      item: elements.map((s) => {
        let author, body, category
        const url = `${domain}${s.website_url || s.canonical_url || ''}`
        const img =
          s.promo_items && (s.promo_items.basic || s.promo_items.lead_art)
        return {
          title: { $: jmespath.search(s, itemTitle) || '' },
          link: url,
          description: { $: jmespath.search(s, itemDescription) || '' },
          guid: {
            '#': url,
            '@isPermaLink': true,
          },
          ...((author = jmespath.search(s, itemCredits)) &&
            author && {
              'dc:creator': { $: author.join(', ') },
            }),
          pubDate: moment
            .utc(s[pubDate])
            .format('ddd, DD MMM YYYY HH:mm:ss ZZ'),
          ...(itemCategory &&
            (category = jmespath.search(s, itemCategory)) &&
            category && { category: category }),
          ...(includeContent !== 0 &&
            (body = flipboardBuildContent.parse(
              s.content_elements || [],
              includeContent,
              domain,
              resizerKey,
              resizerURL,
              resizerWidth,
              resizerHeight,
            )) &&
            body && {
              'content:encoded': {
                $: body,
              },
            }),
          ...(img &&
            img.url && {
              'media:content': {
                '@medium': 'image',
                '@type': 'image/jpeg',
                '@url': buildResizerURL(
                  img.url,
                  resizerKey,
                  resizerURL,
                  resizerWidth,
                  resizerHeight,
                ),
                ...(img.width && {
                  '@width': resizerWidth || img.width,
                }),
                ...(img.height && {
                  '@height': resizerHeight || img.height,
                }),
                ...(jmespath.search(img, imageCaption) && {
                  'media:description': {
                    '@type': 'plain',
                    $: jmespath.search(img, imageCaption),
                  },
                }),
                ...(jmespath.search(img, imageTitle) && {
                  'media:title': {
                    $: jmespath.search(img, imageTitle),
                  },
                }),
                ...((jmespath.search(img, imageCredits) || []).length && {
                  'media:credit': {
                    '@role': 'author',
                    '@scheme': 'urn:ebu',
                    '#': jmespath.search(img, imageCredits).join(','),
                  },
                }),
              },
            }),
        }
      }),
    },
  },
})

export function FlipboardRss({ globalContent, customFields, arcSite }) {
  const {
    resizerURL = '',
    feedDomainURL = '',
    feedTitle = '',
    feedLanguage = '',
  } = getProperties(arcSite)
  const { width = 0, height = 0 } = customFields.resizerKVP || {}

  function FlipboardBuildContent() {
    BuildContent.call(this)

    this.image = (
      element,
      resizerKey,
      resizerURL,
      resizerWidth,
      resizerHeight,
    ) => {
      return {
        figure: {
          img: {
            '@': {
              alt: element.caption || '',
              src: buildResizerURL(
                element.url,
                resizerKey,
                resizerURL,
                resizerWidth,
                resizerHeight,
              ),
              ...(element.width && { width: resizerWidth || element.width }),
              ...(element.height && {
                height: resizerHeight || element.height,
              }),
            },
          },
          ...(element.caption && { figcaption: element.caption }),
        },
      }
    }
  }

  const flipboardBuildContent = new FlipboardBuildContent()

  // can't return null for xml return type, must return valid xml template
  return rssTemplate(get(globalContent, 'content_elements', []), {
    ...customFields,
    resizerURL,
    resizerWidth: width,
    resizerHeight: height,
    domain: feedDomainURL,
    feedTitle,
    feedLanguage,
    flipboardBuildContent,
  })
}

FlipboardRss.propTypes = {
  customFields: PropTypes.shape({
    channelPath: PropTypes.string.tag({
      label: 'Path',
      group: 'Channel',
      description:
        'Path to the feed excluding the domain, defaults to /arc/outboundfeeds/flipboard',
      defaultValue: '/arc/outboundfeeds/flipboard/',
    }),
    ...generatePropsForFeed('rss', PropTypes, ['channelPath', 'includePromo']),
  }),
}

FlipboardRss.label = 'RSS Flipboard'
export default Consumer(FlipboardRss)
