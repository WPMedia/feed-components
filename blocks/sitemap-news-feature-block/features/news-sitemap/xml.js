import PropTypes from 'fusion:prop-types'
import Consumer from 'fusion:consumer'
import getProperties from 'fusion:properties'
import { resizerKey } from 'fusion:environment'
import { BuildPromoItems } from '@wpmedia/feeds-promo-items'
import { generatePropsForFeed } from '@wpmedia/feeds-prop-types'
import jmespath from 'jmespath'

const sitemapTemplate = (
  elements,
  {
    changeFreq,
    includePromo,
    promoItemsJmespath,
    priority,
    lastMod,
    imageTitle,
    imageCaption,
    publicationName,
    newsTitle,
    domain,
    feedTitle,
    newsLanguage,
    resizerURL,
    resizerWidth,
    resizerHeight,
    newsKeywordsJmespath,
    PromoItems,
  },
) => ({
  urlset: {
    '@xmlns': 'http://www.sitemaps.org/schemas/sitemap/0.9',
    '@xmlns:news': 'http://www.google.com/schemas/sitemap-news/0.9',
    ...(includePromo && {
      '@xmlns:image': 'http://www.google.com/schemas/sitemap-image/1.1',
    }),

    url: elements.map((s) => {
      let img
      if (includePromo) {
        img = PromoItems.imageTag({
          ans: s,
          promoItemsJmespath,
          resizerKey,
          resizerURL,
          resizerWidth,
          resizerHeight,
          imageTitle,
          imageCaption,
        })
      }
      const keywords = (jmespath.search(s, newsKeywordsJmespath) || []).join(
        ',',
      )
      const title = newsTitle && jmespath.search(s, newsTitle)

      return {
        loc: `${domain}${s.website_url || s.canonical_url}`,
        ...{ lastmod: s[lastMod] },
        ...(changeFreq !== 'Exclude from sitemap' &&
          changeFreq !== 'Exclude field' && {
            changefreq: changeFreq,
          }),
        ...(priority !== 'Exclude from sitemap' &&
          priority !== 'Exclude field' && { priority: priority }),
        'news:news': {
          'news:publication': {
            'news:name': publicationName || feedTitle,
            ...(newsLanguage &&
              newsLanguage.toLowerCase() !== 'exclude' && {
                'news:language': newsLanguage,
              }),
          },
          'news:publication_date': s[lastMod],
          ...(title && {
            'news:title': { $: title },
          }),
          ...(keywords && {
            'news:keywords': { $: keywords },
          }),
          ...(s.taxonomy &&
            s.taxonomy.stock_symbols && {
              'news:stock_tickers': s.taxonomy.stock_symbols.join(','),
            }),
        },
        ...(img && {
          '#': img,
        }),
      }
    }),
  },
})

export function GoogleSitemap({ globalContent, customFields, arcSite }) {
  const {
    resizerURL = '',
    feedDomainURL = '',
    feedTitle = '',
    feedLanguage = '',
  } = getProperties(arcSite)
  const { width = 0, height = 0 } = customFields.resizerKVP || {}
  const newsLanguage = customFields.newsLanguage || feedLanguage
  const newsKeywordsJmespath =
    customFields.newsKeywords === 'tags'
      ? 'taxonomy.tags[*].text'
      : 'taxonomy.seo_keywords[*]'

  const PromoItems = new BuildPromoItems()

  // can't return null for xml return type, must return valid xml template
  return sitemapTemplate(globalContent.content_elements || [], {
    ...customFields,
    domain: feedDomainURL,
    feedTitle,
    newsLanguage,
    resizerURL,
    resizerWidth: width,
    resizerHeight: height,
    newsKeywordsJmespath,
    PromoItems,
  })
}

GoogleSitemap.propTypes = {
  customFields: PropTypes.shape({
    imageTitle: PropTypes.string.tag({
      label: 'ANS image title key',
      group: 'Featured Media',
      description:
        'ANS value for associated story used for the <image:title> sitemap tag',
      defaultValue: 'title',
    }),
    imageCaption: PropTypes.string.tag({
      label: 'ANS image caption key',
      group: 'Featured Media',
      description:
        'ANS value for associated story image used for the <image:caption> sitemap tag',
      defaultValue: 'caption',
    }),
    publicationName: PropTypes.string.tag({
      label: 'Publication Name',
      group: 'Field Mapping',
      description: 'What name should be used in <news:name> news-sitemap tag',
      defaultValue: '',
    }),
    newsLanguage: PropTypes.string.tag({
      label: 'Language',
      group: 'Field Mapping',
      description:
        'ISO-639 Language code, if blank uses value from feedLanguage in blocks.json. Use Exclude to remove this field.',
      defaultValue: '',
    }),
    newsTitle: PropTypes.string.tag({
      label: 'article title',
      group: 'Field Mapping',
      description: 'Which field should be used from headline',
      defaultValue: 'headlines.basic',
    }),
    newsKeywords: PropTypes.oneOf(['seo_keywords', 'tags']).tag({
      label: 'keywords',
      group: 'Field Mapping',
      description: 'Which field should be used from taxonomy',
      defaultValue: 'seo_keywords',
    }),
    ...generatePropsForFeed('sitemap', PropTypes),
  }),
}
GoogleSitemap.label = 'Sitemap News'
GoogleSitemap.icon = 'browser-page-hierarchy'
export default Consumer(GoogleSitemap)
