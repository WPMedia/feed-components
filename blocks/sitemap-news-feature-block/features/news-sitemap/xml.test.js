// eslint-disable-next-line no-unused-vars
import Consumer from 'fusion:consumer'
import { GoogleSitemap } from './xml'


const articles = {
  content_elements: [
    {
      last_updated_date: '2018-10-05T14:28:25.674Z',
      canonical_url: '/food/2020/04/07/tips-for-safe-hand-washing',
      promo_items: {
        basic: {
          type: 'image',
          title: 'Hand Washing',
          url: 'https://arc-anglerfish-arc2-prod-demo.s3.amazonaws.com/public/STORY_ONE_IMAGE.png',
        },
      },
      headlines: {
        basic:
          'Last Nightâ€™s Match: Tom Brady collects his 500th touchdown pass in the win',
        meta_title: 'dog running outside enjoying the sunny day',
      },
      language: 'en',
      taxonomy: {
        tags: [
          {
            text: 'animal videos',
          },
        ],
      },
    },
    {
      last_updated_date: '2021-11-15T12:18:25.674Z',
      canonical_url: '/food/2021/04/07/best-sourdough-recipes',
      promo_items: {
        mobile: {
          type: 'image',
          title: 'No kneeding around',
          url: 'https://arc-anglerfish-arc2-prod-demo.s3.amazonaws.com/public/STORY_TWO_IMAGE.png',
        },
      },
      headlines: {
        basic: 'Best Sourdough as tested by our staff',
        meta_title: 'Best Sourdough you have ever made',
      },
      language: 'en',
      taxonomy: {
        tags: [
          {
            text: 'food videos',
          },
        ],
      },
    },
  ],
}

it('returns template with default values', () => {
  const googleSitemap = GoogleSitemap({
    arcSite: 'the-globe',
    globalContent: articles,
    customFields: {
      priority: '0.5',
      includePromo: true,
      resizerKVP: {},
      changeFreq: 'always',
      lastMod: 'last_updated_date',
      imageTitle: 'title',
      imageCaption: 'caption',
      newsTitle: 'headlines.meta_title',
      newsKeywords: 'tags',
    },
  })
  expect(googleSitemap).toMatchSnapshot()
})

it('returns template with empty values', () => {
  const googleSitemap = GoogleSitemap({
    arcSite: 'the-globe',
    globalContent: articles,
    customFields: {
      priority: '',
      includePromo: true,
      resizerKVP: {},
      changeFreq: '',
      lastMod: '',
      imageTitle: '',
      imageCaption: '',
      newsTitle: '',
      newsKeywords: '',
      promoItemsJmespath: '',
    },
  })
  expect(googleSitemap).toMatchSnapshot()
})
