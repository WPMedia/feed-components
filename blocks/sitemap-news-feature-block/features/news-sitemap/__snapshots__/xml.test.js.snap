// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns template with default values 1`] = `
Object {
  "urlset": Object {
    "@xmlns": "http://www.sitemaps.org/schemas/sitemap/0.9",
    "@xmlns:image": "http://www.google.com/schemas/sitemap-image/1.1",
    "@xmlns:news": "http://www.google.com/schemas/sitemap-news/0.9",
    "url": Array [
      Object {
        "#": Array [
          Object {
            "image:image": Object {
              "image:loc": "hi/abcdefghijklmnopqrstuvwxyz=/arc-anglerfish-arc2-prod-demo.s3.amazonaws.com/public/STORY_ONE_IMAGE.png",
              "image:title": Object {
                "$": "Hand Washing",
              },
            },
          },
        ],
        "changefreq": "always",
        "lastmod": "2018-10-05T14:28:25.674Z",
        "loc": "http://demo-prod.origin.arcpublishing.com/food/2020/04/07/tips-for-safe-hand-washing",
        "news:news": Object {
          "news:keywords": Object {
            "$": "animal videos",
          },
          "news:publication": Object {
            "news:language": "en",
            "news:name": "google news",
          },
          "news:publication_date": "2018-10-05T14:28:25.674Z",
          "news:title": Object {
            "$": "dog running outside enjoying the sunny day",
          },
        },
        "priority": "0.5",
      },
      Object {
        "changefreq": "always",
        "lastmod": "2021-11-15T12:18:25.674Z",
        "loc": "http://demo-prod.origin.arcpublishing.com/food/2021/04/07/best-sourdough-recipes",
        "news:news": Object {
          "news:keywords": Object {
            "$": "food videos",
          },
          "news:publication": Object {
            "news:language": "en",
            "news:name": "google news",
          },
          "news:publication_date": "2021-11-15T12:18:25.674Z",
          "news:title": Object {
            "$": "Best Sourdough you have ever made",
          },
        },
        "priority": "0.5",
      },
    ],
  },
}
`;

exports[`returns template with empty values 1`] = `
Object {
  "urlset": Object {
    "@xmlns": "http://www.sitemaps.org/schemas/sitemap/0.9",
    "@xmlns:image": "http://www.google.com/schemas/sitemap-image/1.1",
    "@xmlns:news": "http://www.google.com/schemas/sitemap-news/0.9",
    "url": Array [
      Object {
        "changefreq": "",
        "lastmod": undefined,
        "loc": "http://demo-prod.origin.arcpublishing.com/food/2020/04/07/tips-for-safe-hand-washing",
        "news:news": Object {
          "news:publication": Object {
            "news:language": "en",
            "news:name": "google news",
          },
          "news:publication_date": undefined,
        },
        "priority": "",
      },
      Object {
        "changefreq": "",
        "lastmod": undefined,
        "loc": "http://demo-prod.origin.arcpublishing.com/food/2021/04/07/best-sourdough-recipes",
        "news:news": Object {
          "news:publication": Object {
            "news:language": "en",
            "news:name": "google news",
          },
          "news:publication_date": undefined,
        },
        "priority": "",
      },
    ],
  },
}
`;
